homework_1/task_02/taxes/taxes.go

```diff
@@ -0,0 +1,69 @@
+package taxes
+
+import (
+	"github.com/pkg/errors"
+)
+
+type TaxClass struct {
+	UpperAmount  float64
+	Percentage float64
+}
+
+
+func checkLevels(taxClasses []TaxClass) bool {
+	for idx, val := range taxClasses {
+		if idx != len(taxClasses)-1 {
+			if val.UpperAmount >= taxClasses[idx+1].UpperAmount || val.Percentage >= taxClasses[idx+1].Percentage {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func buildTaxClasses(amounts, percentages []float64) []TaxClass{
+	var classes []TaxClass
+
+	for i :=0; i < len(percentages); i++{
+		tempClass := TaxClass{
+			UpperAmount: amounts[i],
+			Percentage: percentages[i],
+		}
+		classes = append(classes, tempClass)
+	}
+	return classes
+}
+
+func CalculateTax(inputValue float64, percentages, amounts []float64) (float64, error){
```

Comment:
Why go through all of the fuss with `buildTaxClasses` when you could simply have this function receive `[]TaxClass` and have it right away?

```suggestion
func CalculateTax(inputValue float64, taxClasses []TaxClass) (float64, error){
```

Reply:
I don't understand how could i have them "right away" if `TaxClass` is a struct, however i do it i need to go through the `buildTaxClasses` to get the `[]TaxClasses`. I can move `buildTaxClasses` to main.go and do it there and then pass the `[]TaxClass` into the `CalculateTax` function. Is this what you meant?

Reply:
First things first - the task is to "implement a library" so everything outside the `homework_1/task_02/taxes` is not part of the library and is actually out-of-the-scope of the task. It's not bad that you implemented a `main` function, it's a good example of how the library can be used.

Now onto your question - what I mean "right away" is that you let the user instantiate the tax classes. You don't even need a helper `buildTaxClasses` function for that:

```
tax, err := taxes.CalculateTax(income, []taxes.TaxClass{
	taxes.TaxClass{
		UpperThreshold: 1000,
		Percentage:     0.0,
	},
	taxes.TaxClass{
		UpperThreshold: 5000,
		Percentage:     0.1,
	},
	taxes.TaxClass{
		UpperThreshold: 10000,
		Percentage:     0.2,
	},
	taxes.TaxClass{
		UpperThreshold: math.Inf(1),
		Percentage:     0.3,
	},
})
```

You can even drop the inner `taxes.TaxClass` since it can be inferred:

```
tax, err := taxes.CalculateTax(income, []taxes.TaxClass{
	{
		UpperThreshold: 1000,
		Percentage:     0.0,
	},
	{
		UpperThreshold: 5000,
		Percentage:     0.1,
	},
	{
		UpperThreshold: 10000,
		Percentage:     0.2,
	},
	{
		UpperThreshold: math.Inf(1),
		Percentage:     0.3,
	},
})
```